applications/FluidDynamicsApplication/custom_processes/two_fluids_inlet_process.cpp

```diff
@@ -0,0 +1,243 @@
+    //          The functions for the distance computation do not work on non-historical variables - this is reason for the following procedure (*)
+
+    // (*) temporally storing the distance field as an older version of itself (it can be assured that nothing is over-written at the start)
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( rRootModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = rRootModelPart.NodesBegin() + i_node;
+        it_node->GetSolutionStepValue(DISTANCE, 2) = it_node->GetSolutionStepValue(DISTANCE, 0);
+    }
+
+    // Preparation for variational distance calculation process
+    // setting distance of inlet nodes to 0.0
+    // setting rest of distances to 1.0
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( rRootModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = rRootModelPart.NodesBegin() + i_node;
+        it_node->GetSolutionStepValue(DISTANCE, 0) = 1.0;
+    }
+
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( mrInletModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = mrInletModelPart.NodesBegin() + i_node;
+        it_node->GetSolutionStepValue(DISTANCE, 0) = -1.0e-7;
+    }
+
+    // Variational distance calculation process is executed to calculate distance from inlet
+    p_dist_proc->Execute();
+
+    // scaling the distance values such that 1.0 is reached at the inlet
+    const double scaling_factor = 1.0 / mInletRadius;
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( rRootModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = rRootModelPart.NodesBegin() + i_node;
+
+        if ( (mInletRadius - it_node->GetSolutionStepValue(DISTANCE, 0)) >= 0 ){
+            // inside the transition radius (from 1 to 0)
+            it_node->GetSolutionStepValue(DISTANCE, 0) = scaling_factor * (mInletRadius - it_node->GetSolutionStepValue(DISTANCE, 0));
+        } else {
+            // outside the transition radius
+            it_node->GetSolutionStepValue(DISTANCE, 0) = 0.0;
+        }
+    }
+
+    // saving the value of DISTANCE to the non-historical variable AUX_DISTANCE
+    VariableUtils var_utils;
+    var_utils.SaveScalarVar( DISTANCE, AUX_DISTANCE, rRootModelPart.Nodes() );
+
+    // (*) restoring the original distance field from its stored version
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( rRootModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = rRootModelPart.NodesBegin() + i_node;
+        it_node->GetSolutionStepValue(DISTANCE, 0) = it_node->GetSolutionStepValue(DISTANCE, 2);
```

Comment:
Same buffer stuff in `it_node->GetSolutionStepValue(DISTANCE, 2)`

Reply:
Done